src/transformStateWithClones.js

```diff
@@ -6,8 +6,34 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        currentState = { ...currentState };
+
+        action.keysToRemove.forEach((key) => {
+          delete currentState[key];
+        });
+
+        break;
+      default:
+        break;
+    }
+    stateHistory.push({ ...currentState });
```

Comment:
[DONT REPEAT YOURSELF] - Pushing a copy of the object to the array is done in every case of the `switch` statement. To adhere to the DRY principle, you should do it just once afterwards, outside of the `switch` block, after the current action is processed.